autopilot/scanner.go

```diff
@@ -2,89 +2,278 @@ package autopilot
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.timings[t.cnt%uint64(len(t.timings))] = timing
+	if t.cnt == math.MaxUint64 {
+		t.cnt = 0
+	} else {
+		t.cnt += 1
+	}
+}
+
+func (t *tracker) timeout() (time.Duration, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if t.cnt < uint64(t.threshold) {
+		return 0, nil
+	}
+
+	percentile, err := stats.Percentile(t.timings, t.percentile)
+	if err != nil {
+		return 0, err
+	}
+
+	return time.Duration(percentile) * time.Millisecond, nil
+}
+
+func newScanner(ap *Autopilot, threads uint8, scanMinInterval, timeoutMinInterval time.Duration) *scanner {
+	s := &scanner{
+		bus:    ap.bus,
+		worker: ap.worker,
+
+		tracker: newTracker(trackerMinDataPoints, trackerNumDataPoints, trackerTimeoutPercentile),
+
+		stopChan: ap.stopChan,
+
+		scanMinInterval:    scanMinInterval,
+		timeoutMinInterval: timeoutMinInterval,
+	}
+	s.pool = newScanPool(s, threads)
+	return s
+}
+
+func (s *scanner) tryPerformHostScan() bool {
 	s.mu.Lock()
-	if s.running {
+	if s.scanning || !s.isScanRequired() {
 		s.mu.Unlock()
-		return
+		return false
 	}
-	s.running = true
+
+	cs, _ := s.bus.ConsensusState()
+	if !cs.Synced {
+		s.mu.Unlock()
+		return false
+	}
+
+	s.scanningLastStart = time.Now()
+	s.scanning = true
 	s.mu.Unlock()
 
 	go func() {
-		s.performHostScan()
+		s.performHostScans()
 		s.mu.Lock()
-		s.running = false
+		s.scanning = false
 		s.mu.Unlock()
 	}()
+	return true
 }
 
-func (s *scanner) performHostScan() {
-	// TODO: initial scan with much shorter
-	// scan up to 10 hosts that we haven't interacted with in at least 1 week
-	hosts, err := s.ap.bus.Hosts(time.Now().Add(-7*24*time.Hour), 10)
-	if err != nil {
-		return
-	}
-	type res struct {
-		hostKey  consensus.PublicKey
-		settings rhpv2.HostSettings
-		err      error
+func (s *scanner) tryUpdateTimeout() time.Duration {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.isTimeoutUpdateRequired() {
+		return s.timeout
 	}
-	resChan := make(chan res)
-	for _, h := range hosts {
-		go func(h hostdb.Host) {
-			scan, err := s.ap.worker.RHPScan(h.PublicKey, h.NetAddress())
-			resChan <- res{h.PublicKey, scan.Settings, err}
-		}(h)
+
+	timeout, _ := s.tracker.timeout() // TODO: handle error
+	if timeout == 0 {
+		return 0
 	}
-	for range hosts {
-		var r res
 
+	s.timeoutLastUpdate = time.Now()
+	s.timeout = timeout
+	return s.timeout
+}
+
+// performHostScans scans every host in our database
+func (s *scanner) performHostScans() {
+	// TODO: running a scan on all hosts is not infinitely scalable, this will
+	// need to be updated to be smarter and run on a subset of all hosts
+	hosts, _ := s.bus.AllHosts()
+	s.pool.launchScans(hosts)
+	inflight := len(hosts)
+
+	// launch workers
+	respChan := s.pool.launchWorkers(func(resChan chan scanResp) {
+		for req := range s.pool.scanQueue {
+			if s.isStopped() {
+				return
+			}
+
+			timeout := s.tryUpdateTimeout()
+			scan, err := s.worker.RHPScan(req.hostKey, req.hostIP, timeout)
```

Comment:
I can't seem to find an implementation so I'm leaving a note here. Is there a sane default timeout in `RHPScan` in case `0` is passed? Otherwise we could spend minutes or hours timing out before we hit the threshold depending on the pool's config. 

Thinking about it a simple constant timeout that a user can configure might also be good enough. Especially since we don't care anyway once a response takes a certain amount of time and if it takes longer than the threshold we can abort and record the max value. But I'll leave that up to @lukechampine to decide.

It just feels a bit nicer to not have the added complexity of checking how much time passed between the last check before we update the timeout again. It also seems a bit more natural than using a percentile. Because if I scan 100 hosts and 99 of them take 10ms and the last one 15ms then it's still great and worth scanning. On the other hand if 99 of them are above 50 seconds and one of them is at 51 seconds none of them might be worth scanning.

Reply:
I am definitely in favour of adding a sane default timeout. I think currently the underlying connection would time out after 5 minutes or so but I will check. With the current configuration the threshold would be reached in a matter of seconds in production, and even if it takes a long time to reach the threshold, we don't pass a duration in that instance which is what we're doing now anyway.

I considered a constant timeout but felt it was a bit overly simple, there's no reason to wait for an arbitrary "long" amount of time if we're not really that interested in the host anyway. What value did you have in mind, David would put it really high for instance.

It only means we're "neglecting" the 5% slowest hosts which shouldn't be that big of a deal, I did consider making it p99 though. I also see your point with the examples but it's probably safe to assume some normal distribution when it comes to host speeds right. At least that's the assumption. I would add a sane upper and lower boundary to the timeout value though in the scanner.